### DISCUSSION TITLE
Controlling Variance with PEP 695 Syntax

### DISCUSSION THREAD

**Discussion by max-muoto** [microsoft/pyright#8433] [Q&A] [answered]
PEP 695 states that traditional type vars are compatible with the new syntax, take this example:
 
 ```python
K = TypeVar("K")


@dataclasses.dataclass
class Container[K: Literal["my_lit", "my_lit_2"]]:
    key: K
```

In this case, I have the desire to manually control the variance, but it would be nice to avoid inheriting from `Generic`.

Take this example, where I want to have less specific literal overloads:

```python
from typing import Generic, Literal, TypeVar, overload
import dataclasses

K = TypeVar("K", covariant=True)


@dataclasses.dataclass
class Container[K: Literal["my_lit", "my_lit_2"]]:
    key: K


@overload
def foo(
    arg_1: Container[Literal["my_lit"]],
) -> None:
    print(arg_1)


@overload
def foo(
    arg_1: Container[Literal["my_lit_2"],],
) -> None:
    print(arg_1)


def foo(arg_1: Container[Literal["my_lit", "my_lit_2"]]) -> None:
    print(arg_1)
```

[Pyright Playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANFACqIkBq5ItYAbmZWOQCYBYAFBYceKH3IxyAY0rkAzgpILhwgNJQAvPUYsQACgBE6o7RmdWScihha6IAK4kAlGqHCAApOlzFyhQA6H1l5JWE-JSgAYTBbclQyAG11AC5CYjIqJKMIOAB9SmIzKFyCoph8gCYjAF1a1OEoZqgAaxI4dPV3L05uXkEhPhJgKGAwMAMmltY0fIBGdNj4xJAkolIKShy8wuL66mEXKABaAD4oADk4kkahFqgEEFQYA1mFtw8vzz6QHn5hMNRuNJtNmu9FjE4tJVutMlsduViNU6tRaochMdzlcbncHk8Xm8QHN5p93ECxhMiSSltCEihkhsstsynsYCVWRUUfUsRdrgy8S0CbZqR9hEA)

Here, K is still treated as invariant, and in order to change that I would have to inherit from `Generic`. [Playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANFACqIkBq5ItYAbmZWOQCa0A4iRRkkAYwCwAKCw48UPuRjlxlcgGcNJDTJkBpKAF56jFiAAUAIn1XaAIzABXFHyNFSFSgG0rEOAD6lMR2UH6BwTABAExWALq04pysSOQoMEZ0IE4kAJR60jIAAkoqapraGgB0parqWjLlWlAAwmDp5KhkFsKiIBLe%2BnG5AFwyUBNQANYkcCNQ%2BgXFnNy8fDJ8JMBQwGBgFuOTrGgBAIzzbR1dIN4eZFS%2B-kEhcQkyuVAAtAB8UABy7RIY2kkygCH66QsxzO%2BUKcKKKxAPH4Gy2Oz2BxBRxAJ3OrXaKmut2I9x84WeUViCTe0g%2BP3%2BgOBoPBqBgUJxMKW0k2212%2B2heMuhL6xM8D3JkVCEuIMXiwy%2BvwBoiZkxZkIFsKAA).

Perhaps this partial compatibility purely exists so that you can reuse the same TypeVar name with old/new syntax, but was curious if this was intended behavior, or not? TypeVars are probably always going to exist in the language, due to the need to tweak variance, but it would be nice to still avoid inheriting from `Generic`.

**Comment by erictraut** [comment, marked as answer]
In your example, the you are defining two symbols `K` in two different scopes. They have no relation to each other.

When you define a class, you need to either use PEP 695 syntax or traditional TypeVars. You cannot combine the two. The PEP is very clear on this.

**Reply by max-muoto** [reply]
Thanks, I appreciate the clarification.
